obspy/clients/eida/fetch.py

```diff
@@ -0,0 +1,950 @@
+                d["end_min"] = "59"
+
+            if int(d["beg_sec"]) > 59:
+                d["beg_sec"] = "59"
+
+            if int(d["end_sec"]) > 59:
+                d["end_sec"] = "59"
+
+            try:
+                beg_time = datetime.datetime(int(d["beg_4year"]),
+                                             int(d["beg_month"]),
+                                             int(d["beg_day"]),
+                                             int(d["beg_hour"]),
+                                             int(d["beg_min"]),
+                                             int(d["beg_sec"]))
+
+                end_time = datetime.datetime(int(d["end_4year"]),
+                                             int(d["end_month"]),
+                                             int(d["end_day"]),
+                                             int(d["end_hour"]),
+                                             int(d["end_min"]),
+                                             int(d["end_sec"]))
+
+            except ValueError as e:
+                self.failstr += "%s [error: wrong begin or end time: %s]\n" \
+                                % (line, str(e))
+                return
+
+            location = "*"
+            cha_list = re.findall("([\w?\*]+)\s*", d["cha_list"])
+
+            if len(cha_list) == int(d['cha_num'])+1:
+                location = cha_list.pop()
+
+            for channel in cha_list:
+                self.postdata += "%s %s %s %s %sZ %sZ\n" \
+                                 % (d["network"],
+                                    d["station"],
+                                    location,
+                                    channel,
+                                    beg_time.isoformat(),
+                                    end_time.isoformat())
+
+        else:
+            self.failstr += "%s [syntax error]\n" % line
+
+    def parse(self, path):
+        with open(path) as fd:
+            for line in fd:
+                if self.__rx_tokenrule.match(line):
+                    continue
+
+                line = line.rstrip()
+
+                if line:
+                    self.__parse_line(line)
+
+
+msglock = threading.Lock()
+
+
+def msg(s, verbose=True):
+    if verbose:
+        with msglock:
+            sys.stderr.write(s + '\n')
+            sys.stderr.flush()
+
+
+def retry(urlopen, url, data, timeout, count, wait, verbose):
+    n = 0
+
+    while True:
+        if n >= count:
+            return urlopen(url, data, timeout)
+
+        try:
+            n += 1
+
+            fd = urlopen(url, data, timeout)
+
+            if fd.getcode() == 200 or fd.getcode() == 204:
+                return fd
+
+            msg("retrying %s (%d) after %d seconds due to HTTP status code %d"
+                % (url, n, wait, fd.getcode()), verbose)
+
+            fd.close()
+            time.sleep(wait)
+
+        except urllib2.HTTPError as e:
+            if e.code >= 400 and e.code < 500:
```

Comment:
This is not granular enough. If I read the code correctly then the `route()` functions is also used for all normal FDSN requests - a lot of `4xx` codes are valid return codes from FDSN servers and thus should not be retried.